pkg/ottl/grammar.go

```diff
@@ -276,8 +276,9 @@ type field struct {
 }
 
 type key struct {
-	String *string `parser:"'[' (@String "`
-	Int    *int64  `parser:"| @Int) ']'"`
+	String     *string          `parser:"'[' (@String "`
+	Int        *int64           `parser:"| @Int"`
+	Expression *mathExprLiteral `parser:"| @@ ) ']'"`
```

Comment:
That's a nice enhancement! thanks for doing it!

Regarding the grammar changes, we need to ensure that new usages of `mathExprLiteral` (or any other visitable node) are being visited by the `grammarVisitor`, otherwise a few things such as custom errors validation and paths extraction wouldn't work for them. 

For example, without adding the extra logic, the statement `set(attributes[append(attributes["tags"], "prod")], "pass")` would be considered valid, as the second editor (`append`) would be "invisible" to the validator.

I think replacing [grammar.go#L346](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/grammar.go#L346) by a custom `path.accept(v)` method that also goes through fields & keys should do the work :)